File: Media Ministry Manager/Forms/Management/ListenersManagement.vb
Imports System.ComponentModel
Imports SPPBC.M3Tools.Events.Listeners

Public Class ListenersManagement

	Private Event ListenerAdded As ListenerEventHandler
	Private Event ListenerDBModified As ListenerEventHandler
	Private Tooled As Boolean = False

	Private Sub Loading(sender As Object, e As EventArgs) Handles Me.Load
		mms_Main.ToggleViewItem("Listeners")
		gt_Email.Authorize(My.Settings.Username)
		gd_Drive.Authorize(My.Settings.Username)
		Reload()
	End Sub

	Private Sub ClosingForm(sender As Object, e As CancelEventArgs) Handles Me.Closing
		' TODO: Find easier way
		If Tooled Then
			Return
		End If

		MainForm.Show()
	End Sub

	Private Sub Logout() Handles mms_Main.Logout
		Helpers.Utils.LogOff()
		Tooled = True
		Me.Close()
	End Sub

	Private Sub ExitApplication() Handles mms_Main.ExitApplication, mms_Main.UpdateAvailable
		Helpers.Utils.CloseOpenForms()
	End Sub

	Private Sub ManageCustomers(sender As Object, e As EventArgs) Handles mms_Main.ManageCustomers
		Dim customers As New CustomersManagement()
		customers.Show()
		Tooled = True
		Me.Close()
	End Sub

	Private Sub ManageOrders(sender As Object, e As EventArgs) Handles mms_Main.ManageOrders
		Dim orders As New OrderManagement()
		orders.Show()
		Tooled = True
		Me.Close()
	End Sub

	Private Sub ManageProducts(sender As Object, e As EventArgs) Handles mms_Main.ManageProducts
		Dim products As New InventoryManagement()
		products.Show()
		Tooled = True
		Me.Close()
	End Sub

	Private Sub ViewSettings() Handles mms_Main.ViewSettings
		Dim settings As New SettingsForm()
		settings.ShowDialog()
	End Sub

	Private Sub SendEmails() Handles dlc_Listeners.SendEmails
		Using emails As New SendEmailsDialog()
			UseWaitCursor = True
			emails.ShowDialog()
			UseWaitCursor = False
		End Using
	End Sub

	Private Sub RemoveListener(sender As Object, e As ListenerEventArgs) Handles dlc_Listeners.RemoveListener
		UseWaitCursor = True
		dbListeners.RemoveListener(e.Listener.Id)
		RaiseEvent ListenerDBModified(Me, e)
	End Sub

	Private Sub AddListener(sender As Object, e As ListenerEventArgs) Handles mms_Main.AddListener, dlc_Listeners.AddListener
		UseWaitCursor = True
		dbListeners.AddListener(e.Listener.Name, e.Listener.Email)
		RaiseEvent ListenerAdded(Me, e)
	End Sub

	Private Sub SendWelcom(sender As Object, e As ListenerEventArgs) Handles Me.ListenerAdded
		UseWaitCursor = True
		Dim subject = "Welcome to the Ministry"
		Dim body = String.Format(newListener, e.Listener.Name.Trim)
		Dim message = gt_Email.Create(e.Listener, subject, body)
#If Not DEBUG Then
		gt_Email.Send(message)
#End If
		RaiseEvent ListenerDBModified(Me, e)
	End Sub

	Private Sub UpdateListener(sender As Object, e As ListenerEventArgs) Handles dlc_Listeners.UpdateListener
		UseWaitCursor = True
		dbListeners.UpdateListener(e.Listener)
		RaiseEvent ListenerDBModified(Me, e)
	End Sub

	Private Sub Reload() Handles dlc_Listeners.RefreshDisplay, Me.ListenerDBModified
		UseWaitCursor = True
		bsListeners.Clear()
		For Each listener In dbListeners.GetListeners()
			bsListeners.Add(listener)
		Next
		dlc_Listeners.Reload()
		UseWaitCursor = False
	End Sub
End Class

File: Media Ministry Manager/Forms/Management/ListenersManagement.cs
using System;
using System.Windows.Forms;

using SPPBC.M3Tools.Events;
using SPPBC.M3Tools.Events.Listeners;

namespace M3App
{
	/// <summary>
	/// Form for managing email listeners
	/// </summary>
	public partial class ListenerManagement
	{
		/// <summary>
		/// 
		/// </summary>
		public ListenerManagement() : base()
		{
			InitializeComponent();

			ldg_Listeners.Reload += new EventHandler(Reload);
			ldg_Listeners.AddListener += new ListenerEventHandler(Add);
			ldg_Listeners.UpdateListener += new ListenerEventHandler(Update);
			ldg_Listeners.RemoveListener += new ListenerEventHandler(Remove);
		}

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		protected override async void Reload(object sender, EventArgs e)
		{
			UseWaitCursor = true;
			_original = await dbListeners.GetListeners();
			ldg_Listeners.Listeners = SPPBC.M3Tools.Types.ListenerCollection.Cast(_original.Items);
			ts_Tools.Count = string.Format(Properties.Resources.COUNT_TEMPLATE, ldg_Listeners.Listeners.Count);
			UseWaitCursor = false;
		}

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		protected override void Remove(object sender, DataEventArgs<SPPBC.M3Tools.Types.Listener> e)
		{
			UseWaitCursor = true;
			_ = dbListeners.RemoveListener(e.Value.Id);
			base.Remove(sender, e);
			Reload(sender, e);
		}

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		protected override void Add(object sender, EventArgs e)
		{
			using SPPBC.M3Tools.Dialogs.AddListenerDialog @add = new();

			if (add.ShowDialog(this) != DialogResult.OK)
			{
				return;
			}

			UseWaitCursor = true;
			AddToDB(sender, add.Listener);
			base.Add(sender, e);
			UseWaitCursor = false;
		}

		/// <inheritdoc/>
		protected override void Import(object sender, EventArgs e)
		{
			using SPPBC.M3Tools.Dialogs.ImportListenersDialog @import = new();

			if (import.ShowDialog(this) != DialogResult.OK)
			{
				return;
			}

			UseWaitCursor = true;
			foreach (SPPBC.M3Tools.Types.Listener listener in import.Listeners)
			{
				AddToDB(sender, listener);
			}

			UseWaitCursor = false;
		}

		private void AddToDB(object sender, SPPBC.M3Tools.Types.Listener listener)
		{
			try
			{
				_ = dbListeners.AddListener(listener);
				SendWelcome(sender, new ListenerEventArgs(listener, EventType.Added));
			}
			catch (Exception ex)
			{
				Console.WriteLine($"Unable to add listener {listener.Name} to database\nError:\n{ex.Message}");
			}
		}

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		protected override void Update(object sender, DataEventArgs<SPPBC.M3Tools.Types.Listener> e)
		{
			try
			{
				UseWaitCursor = true;
				using SPPBC.M3Tools.Dialogs.EditListenerDialog @edit = new(e.Value);

				if (edit.ShowDialog(this) != DialogResult.OK)
				{
					return;
				}

				_ = dbListeners.UpdateListener(edit.Listener);
				base.Update(sender, e);
				Reload(sender, e);
			}
			catch
			{
				_ = MessageBox.Show($"We were unable to update the info for {e.Value.Name}");
			}
			finally
			{
				UseWaitCursor = false;
			}
		}

		/// <inheritdoc/>
		protected override void FilterChanged(object sender, string filter)
		{
			base.FilterChanged(sender, filter);

			ldg_Listeners.Listeners = SPPBC.M3Tools.Types.ListenerCollection.Cast(_original.Items);
		}

		/// <inheritdoc />
		protected override void SendEmails(object sender, EventArgs e)
		{
			UseWaitCursor = true;
			using SendEmailsDialog emails = new();
			_ = emails.ShowDialog(this);
			UseWaitCursor = false;
		}

		private async void SendWelcome(object sender, ListenerEventArgs e)
		{
			UseWaitCursor = true;

#if !DEBUG
			string subject = "Welcome to the Ministry";
			string body = string.Format(Properties.Resources.NEW_LISTENER_TEMPLATE, e.Value.Name.Trim());
			MimeKit.MimeMessage message = gt_Email.Create(e.Value, subject, body);
			await gt_Email.Send(message);
			MessageBox.Show($"Listener {e.Value.Name} has been added to the ministry. Their welcome email should be sent shortly.", "Listener Added", MessageBoxButtons.OK, MessageBoxIcon.Information);
#else
			MessageBox.Show($"Listener {e.Value.Name} has been added to the ministry. Their welcome email should be sent shortly.", "Listener Added", MessageBoxButtons.OK, MessageBoxIcon.Information);
#endif

			Reload(sender, e);
		}
	}
}
